PxGraf/Datasource/FileDatasource/LocalFilesystemDatabase.cs

```diff
@@ -0,0 +1,174 @@
+            IReadOnlyMatrixMetadata meta,
+            IMatrixMap completeTableMap,
+            CancellationToken? cancellationToken = null
+            )
+        {
+            string path = PathUtils.BuildAndSanitizePath(config.DatabaseRootPath, tableReference);
+            if (!path.EndsWith(PxSyntaxConstants.PX_FILE_EXTENSION))
+                path += PxSyntaxConstants.PX_FILE_EXTENSION;
+            DataIndexer indexer = new(completeTableMap, meta);
+            Matrix<DecimalDataValue> output = new(meta, new DecimalDataValue[indexer.DataLength]);
+            using Stream fileStream = File.OpenRead(path);
+            PxFileStreamDataReader dataReader = new(fileStream);
+            if (cancellationToken is null) await dataReader.ReadDecimalDataValuesAsync(output.Data, 0, meta, completeTableMap);
+            else await dataReader.ReadDecimalDataValuesAsync(output.Data, 0, meta, completeTableMap, cancellationToken.Value);
+            return output;
+        }
+
+        /// <inheritdoc/> 
+        public async Task<IReadOnlyMatrixMetadata> GetMatrixMetadataAsync(PxTableReference tableReference)
+        {
+            if (!tableReference.Name.EndsWith(PxSyntaxConstants.PX_FILE_EXTENSION))
+                tableReference.Name += PxSyntaxConstants.PX_FILE_EXTENSION;
+
+            string path = PathUtils.BuildAndSanitizePath(config.DatabaseRootPath, tableReference);
+            using Stream readStream = File.OpenRead(path);
+            PxFileMetadataReader metadataReader = new();
+            IAsyncEnumerable<KeyValuePair<string, string>> entries = metadataReader.ReadMetadataAsync(readStream, config.Encoding);
+            MatrixMetadataBuilder builder = new();
+            MatrixMetadata meta = await builder.BuildAsync(entries);
+            meta = meta.AssignOrdinalDimensionTypes();
+            meta.AssignSourceToContentDimensionValues();
+            return meta;
+        }
+
+        /// <summary>
+        /// Assumes the alias files to be named Alias_[lang].txt where [lang] is the language code of the alias.
+        /// </summary>
+        /// <param name="path">Path to the folder containing the allias files.</param>
+        /// <returns>Multilanguage string of the alias</returns>
+        private MultilanguageString GetGroupName(string path)
+        {
+            Dictionary<string, string> translatedNames = [];
+            IEnumerable<string> aliasFiles = Directory.GetFiles(path, PxSyntaxConstants.ALIAS_FILE_FILTER)
+                .Where(p => Path.GetFileName(p).StartsWith(PxSyntaxConstants.ALIAS_FILE_PREFIX, StringComparison.OrdinalIgnoreCase));
+            foreach (string aliasFile in aliasFiles)
+            {
+                int aliasFileSuffixLength = PxSyntaxConstants.ALIAS_FILE_PREFIX.Length + 1; // +1 for the underscore
+                string lang = new([.. Path.GetFileNameWithoutExtension(aliasFile).Skip(aliasFileSuffixLength)]);
+                Encoding encoding = GetFileEncoding(aliasFile);
```

Comment:
Is it possible to combine these file operations, so only one stream would be needed?

Reply:
Should be possible, changing this